src/main/scala/com/microsoft/hyperspace/actions/RefreshActionBase.scala

```diff
@@ -76,4 +78,56 @@ private[actions] abstract class RefreshActionBase(
           s"Current index state is ${previousIndexLogEntry.state}")
     }
   }
+
+  /**
+   * Compare list of source data files from previous IndexLogEntry to list
+   * of current source data files, validate fileInfo for existing files and
+   * identify deleted source data files.
+   */
+  protected lazy val deletedFiles: Seq[String] = {
+    val rels = previousIndexLogEntry.relations
+    val originalFiles = rels.head.data.properties.content.fileInfos
+    val currentFiles = rels.head.rootPaths
+      .flatMap { p =>
+        Content
+          .fromDirectory(path = new Path(p), throwIfNotExists = true)
+          .fileInfos
+      }
+      .map(f => f.name -> f)
+      .toMap
+
+    var delFiles = Seq[String]()
+    originalFiles.foreach { f =>
+      currentFiles.get(f.name) match {
+        case Some(v) =>
+          if (!f.equals(v)) {
+            throw HyperspaceException(
+              "Index refresh (to handle deleted source data) aborted. " +
+                s"Existing source data file info is changed (file: ${f.name}).")
+          }
+        case None => delFiles :+= f.name
+      }
+    }
+
+    delFiles ++ previousIndexLogEntry.deletedFiles
+  }
+
+  protected lazy val appendedFiles = {
+    val relation = previousIndexLogEntry.relations.head
+
+    // TODO: improve this to take last modified time of files into account.
+    //   https://github.com/microsoft/hyperspace/issues/182
+    val originalFiles = relation.data.properties.content.files.map(_.toString)
```

Comment:
Can we handle this issue with this PR?
It might cause some correctness issue:
- modified files are handled as deleted => delete the index rows
- not handled as appended => missing data

Also I think we should consider the following scenario:
- index creation without linage column
- some of source files are modified
- refresh incremental mode:
  - delete => failed, do nothing
  - append => should throw an exception

Reply:
thanks @sezruby , I actually prefer not to handle #182 in this PR.
we know it's a correctness bug if user replaces a data file with another file with exact same name, but I think that will be solved when we handle #182 (soon).

I will keep this issue open for the other scenario you pointed out.